mongo/integration/csot_test.go

```diff
@@ -455,6 +455,46 @@ func TestCSOT(t *testing.T) {
 			bsoncore.ErrElementNotFound,
 			"expected maxTimeMS BSON value to be missing, but is present")
 	})
+
+	// TODO(GODRIVER-2944): Remove this test once the "timeoutMode" option is
+	// supported.
+	mt.RunOpts("Find always sends manually-specified maxTimeMS", csotOpts, func(mt *mtest.T) {
```

Comment:
This test is being run with a client-level timeout of 10 seconds, set by previous subtests. The csotOpts should be reset before executing to test that the deprecated operation-level maxTimeMS value is being propagated to command message when omitting CSOT.

```suggestion
	csotOpts = mtest.NewOptions().ClientOptions(options.Client())
	mt.RunOpts("Find always sends manually-specified maxTimeMS", csotOpts, func(mt *mtest.T) {
```

Suggest adding the following function to test that a 5s timeout is being used:

```go
assertMaxTimeMS := func(mt *mtest.T, command bson.Raw, expected int64) {
	mt.Helper()

	maxTimeVal := command.Lookup("maxTimeMS")

	require.Greater(mt,
		len(maxTimeVal.Value),
		0,
		"expected maxTimeMS BSON value to be non-empty")
	require.Equal(mt,
		maxTimeVal.Type,
		bson.TypeInt64,
		"expected maxTimeMS BSON value to be type Int64")
	assert.Greater(mt,
		maxTimeVal.Int64(),
		int64(0),
		"expected maxTimeMS value to be greater than 0")
	assert.Equal(mt, expected, maxTimeVal.Int64())
}
```

Also suggest creating an additional test ensuring behavior defined in the [Validation and Overrides](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.md#validation-and-overrides) section of the CSOT specifications:

> When executing an operation, drivers MUST ignore any deprecated timeout options if timeoutMS is set on the operation or is inherited from the collection/database/client levels.


```go
csotOpts = mtest.NewOptions().ClientOptions(options.Client().SetTimeout(15 * time.Second))
mt.RunOpts("Find sends client-level timeout over deprecated maxTimeMS", csotOpts, func(mt *mtest.T) {
	ctx := context.Background()

	// Manually set a maxTimeMS on the Find and assert that it's sent with
	// the command, even when CSOT is enabled.
	cursor, err := mt.Coll.Find(
		ctx,
		bson.D{},
		options.Find().SetMaxTime(5*time.Second))
	require.NoError(mt, err, "Find error")
	err = cursor.Close(context.Background())
	require.NoError(mt, err, "Cursor.Close error")

	evt := getStartedEvent(mt, "find")
	fmt.Println(evt)
	//assertMaxTimeMSIsSet(mt, evt.Command)
	assertMaxTimeMS(mt, evt.Command, 15000)
})
```

Reply:
I think the test matrix should look something like this:

1. no client-level, operation-level, deprecated operation-level -> use DOL
2. ~CL, ~OL, DOL -> DOL
3. CL, OL, DOL -> DOL
4. CL, ~OL, DOL -> CL

Reply:
The current test case is basically
> CL, OL, DOL -> DOL
so setting `timeoutMS` is intentional. However, the test doesn't cover the rest of the cases, which could be confusing (especially because the test description is actually incorrect, since a manually-set `maxTimeMS` is ignored when a client-level timeout is set but no operation-level timeout is set).

I'll update the test to cover the additional cases.

Reply:
I added test cases for the permutations you described.